spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -585,6 +625,54 @@ private void ackImmediate(ConsumerRecord<K, V> record) {
 		}
 
 		private void invokeListener(final ConsumerRecords<K, V> records) {
+			if (this.isBatchListener) {
+				invokeBatchListener(records);
+			}
+			else {
+				invokeRecordListener(records);
+			}
+		}
+
+		private void invokeBatchListener(final ConsumerRecords<K, V> records) {
+			List<ConsumerRecord<K, V>> recordList = new LinkedList<ConsumerRecord<K, V>>();
+			Iterator<ConsumerRecord<K, V>> iterator = records.iterator();
+			while (iterator.hasNext()) {
+				recordList.add(iterator.next());
+			}
+			if (recordList.size() > 0) {
+				try {
+					if (this.batchAcknowledgingMessageListener != null) {
+						this.batchAcknowledgingMessageListener.onMessage(recordList,
+								new ConsumerBatchAcknowledgment(recordList, this.isManualImmediateAck));
+					}
+					else {
+						this.batchListener.onMessage(recordList);
+						if (!this.isAnyManualAck && !this.autoCommit) {
+							for (ConsumerRecord<K, V> record : recordList) {
+								this.acks.put(record);
+							}
+						}
+					}
+				}
+				catch (Exception e) {
+					if (this.containerProperties.isAckOnError() && !this.autoCommit) {
+						for (ConsumerRecord<K, V> record : recordList) {
+							this.acks.add(record);
```

Comment:
Both `put` and `add` can throw an `IllegalArgumentException` - it's based on the argument, not the queue state. If we change this to `put` we'd have to add an unnecessary try/catch here - it's un unbounded queue.